weave/trace_server/execution_runner/cross_process_trace_server.py

```diff
@@ -0,0 +1,540 @@
+
+        # Flag to track if we're shutting down - MUST be set before starting thread
+        self._shutdown = False
+
+        # Start response handler thread
+        self._response_handler_thread = threading.Thread(
+            target=self._handle_responses, daemon=True
+        )
+        self._response_handler_thread.start()
+
+    def _generate_request_id(self) -> str:
+        """Generate a unique request ID."""
+        with self._lock:
+            self._request_counter += 1
+            return f"req_{self._request_counter}"
+
+    def _handle_responses(self) -> None:
+        """
+        Handle responses from the worker thread.
+
+        This method runs in a separate thread and continuously polls the
+        response queue for incoming responses, matching them with pending
+        requests.
+        """
+        while not self._shutdown:
+            try:
+                response = self._response_queue.get(timeout=0.1)
+                request_id = response["request_id"]
+
+                with self._lock:
+                    if request_id in self._pending_requests:
+                        self._responses[request_id] = response
+                        self._pending_requests[request_id].set()
+            except queue.Empty:
+                continue
+            except Exception as e:
+                # Log error but continue handling responses
+                logger.exception(f"Error handling response: {e}")
+
+    def _send_request(self, method: str, request: Any) -> Any:
+        """
+        Send a request to the worker thread and wait for the response.
+
+        Args:
+            method: The method name to call
+            request: The request object
+
+        Returns:
+            The response from the worker thread
+
+        Raises:
+            RuntimeError: If the server has been shut down
+            TimeoutError: If the request times out
+            SendRequestException: If an error occurs in the worker thread
+        """
+        if self._shutdown:
+            raise RuntimeError("CrossProcessTraceServer has been shut down")
+
+        request_id = self._generate_request_id()
+        event = threading.Event()
+
+        with self._lock:
+            self._pending_requests[request_id] = event
+
+        # Send the request
+        wrapped_request = RequestWrapper(
+            method=method, request=request, request_id=request_id
+        )
+        self._request_queue.put(wrapped_request)
+
+        # Wait for the response
+        if not event.wait(timeout=TIMEOUT_SECONDS):
+            raise TimeoutError(f"Request {request_id} timed out")
+
+        # Get the response
+        with self._lock:
+            response = self._responses.pop(request_id)
+            del self._pending_requests[request_id]
+
+        if response["error"]:
+            raise SendRequestException(f"Worker thread error: {response['error']}")
+
+        return response["result"]
+
+    def shutdown(self) -> None:
```

Comment:
does this get called automatically on cleanup? It's not immediately clear under what conditions the cross process trace server gets cleaned up. Like, when the `_run_user_scoped_function` finishes?